### PR TITLE
HomeWork #6

### PR DESCRIPTION
**Pull request by Timurovic**
## Description
Use MVVM for Movies and MoviewDetails screens

## Changes
- Add ViewModels and Livedata to project
- Moved load data to ViewModels

### CODE REVIEW

File: app/src/main/java/com/android/academy/fundamentals/homework/features/moviedetails/MovieDetailsFragment.kt [hunk 1]
```diff
@@ -59,27 +54,28 @@ class MovieDetailsFragment : Fragment() {
...
 
-            if (movie != null) {
-                bindUI(view, movie)
-            } else {
-                showMovieNotFoundError()
-            }
+        viewModel = MovieDetailsViewModel((requireActivity() as MovieRepositoryProvider).provideMovieRepository())
+        viewModel.loadDetails(movieId)
+
+        lifecycleScope.launch {
```
Comment (ermolnik) [inline app/src/main/java/com/android/academy/fundamentals/homework/features/moviedetails/MovieDetailsFragment.kt:61]: А зачем liveData оборачивать в корутину?

Comment (Timurovic) [reply app/src/main/java/com/android/academy/fundamentals/homework/features/moviedetails/MovieDetailsFragment.kt:61]: как раз этот кусок меня смущает)

File: app/src/main/java/com/android/academy/fundamentals/homework/features/moviedetails/MovieDetailsFragment.kt [hunk 2]
```diff
@@ -22,9 +22,9 @@ import com.android.academy.fundamentals.homework.di.MovieRepositoryProvider
 import com.android.academy.fundamentals.homework.model.Movie
 import kotlinx.coroutines.launch
 
-
 class MovieDetailsFragment : Fragment() {
 
+    private lateinit var viewModel :MovieDetailsViewModel
```
Comment (emartynov) [inline app/src/main/java/com/android/academy/fundamentals/homework/features/moviedetails/MovieDetailsFragment.kt:27]: Formatting?

File: app/src/main/java/com/android/academy/fundamentals/homework/features/moviedetails/MovieDetailsFragment.kt [hunk 3]
```diff
@@ -59,27 +54,26 @@ class MovieDetailsFragment : Fragment() {
...
-            val movie = repository.loadMovie(movieId)
-
-            if (movie != null) {
-                bindUI(view, movie)
-            } else {
-                showMovieNotFoundError()
-            }
-        }
+
+        viewModel = MovieDetailsViewModel((requireActivity() as MovieRepositoryProvider).provideMovieRepository())
```
Comment (emartynov) [inline app/src/main/java/com/android/academy/fundamentals/homework/features/moviedetails/MovieDetailsFragment.kt:58]: ViewModelProvider?

File: app/src/main/java/com/android/academy/fundamentals/homework/features/moviedetails/MovieDetailsFragment.kt (hunk 2 above)
Comment (emartynov) [inline app/src/main/java/com/android/academy/fundamentals/homework/features/moviedetails/MovieDetailsFragment.kt:27]: Do you need to keep it in the class field?

File: app/src/main/java/com/android/academy/fundamentals/homework/features/moviedetails/MovieDetailsViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.android.academy.fundamentals.homework.data.MovieRepository
+import com.android.academy.fundamentals.homework.model.Movie
+import kotlinx.coroutines.launch
+
+class MovieDetailsViewModel(private val repository: MovieRepository) : ViewModel() {
+
+    private val _movie = MutableLiveData<Movie?>()
```
Comment (emartynov) [inline app/src/main/java/com/android/academy/fundamentals/homework/features/moviedetails/MovieDetailsViewModel.kt:13]: Why nullable?

File: app/src/main/java/com/android/academy/fundamentals/homework/features/movies/MoviesListFragment.kt [hunk 5]
```diff
@@ -6,17 +6,15 @@ import android.view.LayoutInflater
...
 import androidx.recyclerview.widget.RecyclerView
 import com.android.academy.fundamentals.homework.R
 import com.android.academy.fundamentals.homework.di.MovieRepositoryProvider
 import com.android.academy.fundamentals.homework.model.Movie
-import kotlinx.coroutines.launch
-
 
 class MoviesListFragment : Fragment() {
 
+    private lateinit var viewModel: MoviesListViewModel
```
Comment (emartynov) [inline app/src/main/java/com/android/academy/fundamentals/homework/features/movies/MoviesListFragment.kt:17]: Same questions about filed, view model provider

File: app/src/main/java/com/android/academy/fundamentals/homework/features/movies/MoviesListViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.android.academy.fundamentals.homework.data.MovieRepository
+import com.android.academy.fundamentals.homework.model.Movie
+import kotlinx.coroutines.launch
+
+class MoviesListViewModel(private val repository: MovieRepository) : ViewModel() {
+
+    private val _movies = MutableLiveData<List<Movie>>(emptyList())
```
Comment (emartynov) [inline app/src/main/java/com/android/academy/fundamentals/homework/features/movies/MoviesListViewModel.kt:13]: Why empty list?

**Review by emartynov — COMMENTED** [review]
Couple of comments

File: app/src/main/java/com/android/academy/fundamentals/homework/features/moviedetails/MovieDetailsFragment.kt (hunk 3 above)
Comment (Timurovic) [reply app/src/main/java/com/android/academy/fundamentals/homework/features/moviedetails/MovieDetailsFragment.kt:58]: as I know ViewModelProvider create ViewModel without passing parameters/arguments

Comment (emartynov) [reply app/src/main/java/com/android/academy/fundamentals/homework/features/moviedetails/MovieDetailsFragment.kt:58]: Yup, you have to provide a Factory for ViewModel creation

File: app/src/main/java/com/android/academy/fundamentals/homework/features/moviedetails/MovieDetailsViewModel.kt (hunk 4 above)
Comment (Timurovic) [reply app/src/main/java/com/android/academy/fundamentals/homework/features/moviedetails/MovieDetailsViewModel.kt:13]: repository returned nullable object, so I didn't refactored it

File: app/src/main/java/com/android/academy/fundamentals/homework/features/movies/MoviesListViewModel.kt (hunk 6 above)
Comment (Timurovic) [reply app/src/main/java/com/android/academy/fundamentals/homework/features/movies/MoviesListViewModel.kt:13]: my mistake

File: app/src/main/java/com/android/academy/fundamentals/homework/features/moviedetails/MovieDetailsFragment.kt (hunk 2 above)
Comment (Timurovic) [reply app/src/main/java/com/android/academy/fundamentals/homework/features/moviedetails/MovieDetailsFragment.kt]: why not? 
viewModel used only into this fragment

**Review by emartynov — APPROVED** [review]
LGTM

**Pull request merged**